### ISSUE TITLE
"Teaching Elm to Beginners" by Richard Feldman

### ISSUE THREAD

**Issue by nelsonic** [dwyl/learn-elm#97]
[![image](https://user-images.githubusercontent.com/194400/31385240-68feeb9c-adba-11e7-93a7-65d378698403.png)](https://youtu.be/G-GhUxeYc1U)
https://youtu.be/G-GhUxeYc1U

**Comment by YvesMuyaBenda** [comment]
@nelsonic The man is on point! But come on! Swbat is just user-stories for teachers ;) Anyways, my main way of looking at information like this is to (1) ask how does this information help in the writing of technial texts (writing out is a form of self-teaching) and (2) how can this be adapted to become a more efficientt self-learner (stuff like swbat is useful for scoping out self-learning projects). Anyways it was fun to have a different angle to the topic; it helped me clarify a few things about consciously designing learning.

**Comment by nelsonic** [comment]
Yeah, it's always good to get a different angle. Richard has _taught_ Elm to more people than _anyone_ `else` so he has the _most_ insight into the most common "_beginner_" questions.
I'm finding it _especially_ useful as I'm trying to teach Java Devs how to a) write front-end code and b) write `functional` front-end code!! 😉

**Comment by YvesMuyaBenda** [comment]
@nelsonic You are too hard-core for me!

**Comment by YvesMuyaBenda** [comment]
There is a basic principle that he does not explicitly state which is often used in teaching programming concepts: learning by refactoring. This is where one refactors (rewriting code while keeping some defintiion of external behaviour) code using a new concept based upon a perceived problem-solution pair; e.g. his explanation of union types: "imagine if we added 20 more things ..." It can be used at any level: hey, do you see how these expressions are similar but slightly different and they all solve the same underlying problem? Yay, functions, functions sort that out mate! Write some functions! This is what abstraction is all about, etc, etc!

**Comment by YvesMuyaBenda** [comment]
So just thinking out loud (and my knowledge of Java is not even cursory) applying the above thoughts, you might start with a GUI Java app in all of its full OOP glory, and rebuild it on the front end solving the problems from a front-end FP point of view, while comparing how similar problems find different solutions in different paradigms. Heck the app could be a simple counter, and there you might start with how OOP and FP model differently, not object with methods which are instantiated from classes (data + functionally being binded together in one concept with objects being abstraed within classes), but data being transformed through a pipline (one essetially models the concepts algebraically, and there is no need as yet to have an abstract counter, one could just model ith with an integer). I am just thinking out loud, it would be an excellent exercise to spell out the details and it is similar to what you did with the tutorioal in the Elm architecure but .. My main thought is a simple easily understood problem which allows you to take a nice slice out of the necessary programming concepts is the simplest approach I can think of. Anyways that was fun!

**Comment by YvesMuyaBenda** [comment]
Yet more thoughts! For me, it all hinges upon the concept of state and how different paradigms form different interpretations of how it should be modeled/implemented; but I have only seen a handful of places (including the issue on the 'Computational Thinking in Elm' book in this repo) that even start talking about it coherently.  State is a concept that would have to come really early methinks.
